docs/proposals/manifest-hydrator.md

```diff
@@ -0,0 +1,538 @@
+
+If output is not deterministic, then a preview generated today might not be valid/correct a week, day, or even hour later. Non-determinism makes it so that developers can't trust that the change they review will be the change actually applied.
+
+### Non-Goals
+
+1) Implementing a change promotion system
+
+## Open Questions
+
+* The `sourceHydrator` field is mutually exclusive with the `source` and the `sources` field. Should we throw an error if they're both configured, or should we just pick one and ignore the others?
+* How will/should this feature relate to the image updater? Is there an opportunity to share code, since both tools involve pushing to git?
+* Should we enforce a naming convention for hydrated manifest branches, e.g. `argo/...`? This would make it easier to recommend branch protection rules, for example, only allow pushes to `argo/*` from the argo bot.
+* Should we enforce setting a `sourceHydrator.syncSource.path` to something besides `.`? Setting a path makes it easier to add/remove other apps later if desired.
+
+## Proposal
+
+Today, Argo CD watches one or more git repositories (configured in the `spec.source` or `spec.sources` field). When a new commit appears, Argo CD updates the desired state by rendering the manifests with the configured manifest hydration tool. If auto-sync is enabled, Argo CD applies the new manifests to the cluster.
+
+With the introduction of this change, Argo CD will watch two revisions in the same git repository: the first is the "dry source", i.e. the git repo/revision where the un-rendered manifests reside, and the second is the "hydrated source," where the rendered manifests are places and retrieved for syncing to the cluster.
+
+### New `spec.sourceHydrator` Application Field
+
+A `sourceHydrator` field will be added to the Argo CD Application spec:
+
+```yaml
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: example
+spec:
+  # The sourceHydrator field is mutually-exclusive with `source` and with `sources`. If this field is configured, we 
+  # should either throw an error or ignore the other two.
+  sourceHydrator:
+    drySource:
+      repoURL: https://github.com/argoproj/argocd-example-apps
+      targetRevision: main
+      # This assumes the Application's environments are modeled as directories.
+      path: environments/e2e
+    syncSource:
+      targetBranch: environments/e2e
+      path: .
+    # The hydrateTo field is optional. If specified, Argo CD will write hydrated manifests to this branch instead of the
+    # syncSource.targetBranch. This allows the user to "stage" a hydrated commit before actually deploying the changes
+    # by merging them into the syncSource branch. A complete change promotion system can be built around this feature. 
+    hydrateTo:
+      targetBranch: environments/e2e-next
+      # The path is assumed to be the same as that in syncSource.
```

Comment:
Echo-ing @jessesuen a little bit here but have additional thoughts so I figure I'd start another thread on this.

The more and more I think about this the more I am leaning towards it being a separate CRD. ApplicationSets set the precedent. It's a separate controller with it's own CRD that had the scope of "just an Application generating factory". Similarly, this can be thought of in the same light.

From my PoV, I can think of two things...

1. Have a `Hydrator` CRD with the same `drySource` and `hydrateTo` spec but the end result is that the Hydrator Controller generates an Argo CD Application using the `hydrateTo` as it's `source`
2. (^ related but also separate thought) If going down this path, maybe just make it another ApplicationSet generator? "Hydrator Generator" with the scope of it bing "a generator that just generates one App".

Just my brainstorming thoughts.

Reply:
Will respond to Jesse ASAP!

ApplicationSet has a 1:N relationship with Applications, so it fundamentally can't be embedded in a single app. It's a fundamental aspect of the tool, which is why it required a separate CRD. But using a separate CRD has had numerous (unavoidable) downsides: mainly that we've had to introduce new APIs and RBAC to provide basic first-class support and that we still don't have full UI/CLI support because dealing with a completely new CRD comes with a lot of overhead.

I believe source hydration should have a 1:1 relationship with Applications, because it's fundamentally something that an individual developer cares about on an individual app level. If I'm a developer maintaining my one app in a big organization, I should have the ability to easily use the rendered manifest pattern without going to an Ops team to configure some separate resource for me.

As a feature with a 1:1 relationship to an Application, I think hydration is similar to multi-source apps: it's an opt-in feature, but if you want to use it, it's configured the same place as all your other app config.

> Have a Hydrator CRD with the same drySource and hydrateTo spec but the end result is that the Hydrator Controller generates an Argo CD Application using the hydrateTo as it's source

This suffers from all the problems of introducing _any_ new CRD, but it also has its own problems. This is what the spec would probably look like:

```yaml
kind: HydratedApplication
spec:
  drySource:
    repoURL:
    targetRevision:
    path:
  syncSource:
    targetBranch:
    path:
  specTemplate:
    # ... insert every Argo CD field except `source` and `sources`
```

This weirdly gives the controller two completely separate tasks: 1) do manifest hydration, and 2) don't let the user set the `source` or `sources` field. So half of the controller's job becomes this unusual "enforcement" task. And users will encounter all the pain points they encounter with ApplicationSets: they don't know why their UI changes keep getting overwritten suddenly, and they don't have easy access to modify the higher-level (hydrator) config.

> (^ related but also separate thought) If going down this path, maybe just make it another ApplicationSet generator? "Hydrator Generator" with the scope of it bing "a generator that just generates one App".

I think this suffers from the same problems as the `HydratedApplication` and also introduces a strange new task for the ApplicationSet controller of hydrating manifests and pushing them to git.

Reply:
Quick follow-up: when we considered multi-source support, we also considered whether you could do the same thing via ApplicationSet. In that case, it's technically possible, and even fits within AppSet's existing functionality (just use a git files generator and template the values file into `sources.helm.values`).

We decided that having a first-class UX via the Application spec was a better move than using the AppSet approach.

Reply:
I have used ApplicationSet heavily and I don't think the two feature are comparable for the same reasons explained above. 

In all the features of argo, multi-source seems like the closest comparable feature: they both impact what comes in and what goes out, they both have different behavior/spec related to hydration. 

On the API/CRD level, they both bring the same concerns: 
- when `sources` is used, `source` cannot. => when `sourceHydrator` is used, `sources` and `source` cannot.
- UI change were required for the UI, and it was done iteratively through the versions without impact for non-multisource users 
- CLI changes were required and it was done iteratively
- and so on.

I think the main difference here, is the technical complexity to implement the feature. The implementation of source hydration shouldn't have a negative effect on the current controllers. Most of the logic for source hydration would benefit from being executed in other micro-services, and not in the application-controller - which is already complex today - as much as possible.

But from an API design point-of-view, I think it makes sense to have these configuration as part of the Application natively. Now I think we should give more thoughts to these configuration, and I will start another thread for those like @christianh814 did 😄